added functionality to receive Events from Openhab and added methods to create and delete openhab items

galexey opened the pull request:
I also created som testcases for the new functionality.

galexey committed:
initial commit for added functionality to receive ItemEvents from Openhab

galexey committed:
receiving events works.
need write testcases still

galexey committed:
ItemFactory to create new items
create item test

galexey committed:
added testcases for creation of all itemtypes
basic tests for setting states and commands
added a method to delete items

galexey committed:
added testcases
finished documentation

sim0nx commented:
Hi @galexey 
Thanks for your PR!
That is quite a big one to review though ;-)
Could you please make sure to rebase from master and cleanup your code according to pylint, flake8 and mypy as much as possible.

What are you thoughts on using aiohttp-sse instead of threading+sseclient?

galexey committed:
cleaning code (PEPs)
renaming of literals to python style

galexey committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	openhab/client.py
#	openhab/items.py

galexey committed:
merged with upstream master
cleaning code regarding new codestyles

galexey commented:
Hi!
Jep, this one grew quite big, sorry.
I merged your changes and did some major cleanup (also applied python naming style)
I haven´t used aiohttp-sse yet, but it looks nice. I might get back to your proposal and have a deeper look into it later this month.

galexey committed:
updated samples in readme

galexey committed:
fixed bug in event-parsing and testcases

galexey commented:
I need to rework Commandhandling a bit. expect another commit soon.

galexey committed:
major refactor to better reflect incoming states/statechange and commands.

changed handling of incoming values

wrote quite a lot of testcases for creation/deletion of items

wrote quite a lot of testcases for event handling

sim0nx commented on openhab/client.py:
```diff
@@ -143,7 +375,7 @@ def fetch_all_items(self) -> typing.Dict[str, openhab.items.Item]:
       dict: Returns a dict with item names as key and item class instances as value.
     """
-    items = {}  # type: dict
+    items = {}  # data_type: dict
```
this should really be type

sim0nx commented on openhab/items.py:
```diff
@@ -19,26 +19,183 @@
   """Base item class."""
 
-  types = []  # type: typing.List[typing.Type[openhab.types.CommandType]]
+  types = []  # data_type: typing.List[typing.Type[openhab.types.CommandType]]
```
this should really be type

sim0nx commented on openhab/items.py:
```diff
@@ -47,16 +204,33 @@ def __init__(self, openhab_conn: 'openhab.client.OpenHAB', json_data: dict) -> N
-    self._state = None  # type: typing.Optional[typing.Any]
-    self._raw_state = None  # type: typing.Optional[typing.Any]  # raw state as returned by the server
-    self._members = {}  # type: typing.Dict[str, typing.Any] #  group members (key = item name), for none-group items it's empty
+    self._state = None  # data_type: typing.Optional[typing.Any]
```
this should really be type

sim0nx commented on openhab/items.py:
```diff
@@ -47,16 +204,33 @@ def __init__(self, openhab_conn: 'openhab.client.OpenHAB', json_data: dict) -> N
-    self._raw_state = None  # type: typing.Optional[typing.Any]  # raw state as returned by the server
-    self._members = {}  # type: typing.Dict[str, typing.Any] #  group members (key = item name), for none-group items it's empty
+    self._state = None  # data_type: typing.Optional[typing.Any]
+    self._raw_state = None  # data_type: typing.Optional[typing.Any]  # raw state as returned by the server
```
this should really be type

sim0nx commented on openhab/items.py:
```diff
@@ -47,16 +204,33 @@ def __init__(self, openhab_conn: 'openhab.client.OpenHAB', json_data: dict) -> N
-    self._members = {}  # type: typing.Dict[str, typing.Any] #  group members (key = item name), for none-group items it's empty
+    self._state = None  # data_type: typing.Optional[typing.Any]
+    self._raw_state = None  # data_type: typing.Optional[typing.Any]  # raw state as returned by the server
+    self._raw_state_event = None  # data_type: str  # raw state as received from Serverevent
```
this should really be type

sim0nx commented on openhab/items.py:
```diff
@@ -47,16 +204,33 @@ def __init__(self, openhab_conn: 'openhab.client.OpenHAB', json_data: dict) -> N
+    self._state = None  # data_type: typing.Optional[typing.Any]
+    self._raw_state = None  # data_type: typing.Optional[typing.Any]  # raw state as returned by the server
+    self._raw_state_event = None  # data_type: str  # raw state as received from Serverevent
+    self._members = {}  # data_type: typing.Dict[str, typing.Any] #  group members (key = item name), for none-group items it's empty
```
this should really be type

sim0nx commented on openhab/items.py:
```diff
@@ -128,13 +329,16 @@ def _validate_value(self, value: typing.Union[str, typing.Type[openhab.types.Com
   def _rest_format(self, value: str) -> typing.Union[str, bytes]:
     """Format a value before submitting to openHAB."""
-    _value = value  # type: typing.Union[str, bytes]
+    _value = value  # data_type: typing.Union[str, bytes]
```
this should really be type

sim0nx commented on openhab/client.py:
```diff
@@ -88,6 +106,186 @@ def _check_req_return(req: requests.Response) -> None:
+
+
+
+        # event = None
```
no commented code blocks

sim0nx commented on openhab/client.py:
```diff
@@ -88,6 +106,186 @@ def _check_req_return(req: requests.Response) -> None:
+        else:
+          self.logger.debug("item '{}' not registered. ignoring the arrived event.".format(item_name))
+
+
```
too many newlines

sim0nx commented:
A couple of change requests:
- please do fix linting issues (too many newlines, no newline at end of file etc)
- do not use data_type in docstrings. either *data type* or just leave *type* please
- could you please strip the SSE stuff into a separate pull request so that we can differentiate between code reworks, new type support, more complete API support ... and the event stream handling ?
- please also fix the tests which currently fail ... there are at least import errors

Thanks

galexey committed:
changed Server Side Events client to aiohttp_sse_client as suggested by Georges
some additions around event content
bugfixes
cleaned code to fix linting issues

galexey commented:
Hi!
I cleaned the code as much as I cound and found reasonable.
the suggested aiohttp_sse_client works quite well and is a cleaner approach - thx for the hint.

However I can not split up the PRs to isolate the SSE-stuff, I had to change a lot of things in client, items and types modules for sse-implementation.

galexey committed:
added http_buffersize to avoid "line too long" Exception
when reading images from OH

galexey committed:
added audiosinks
added voices
added voice interpreter

added cache for items

sim0nx commented:
Hi,
I cloned your changes into a local branch and spend some time cleaning things up a bit, namely docstrings and typing.

Could you please rebase your PR on that branch ( https://github.com/sim0nx/python-openhab/tree/galexey_pr ) ... if required open a new PR.
I think it is important to have cleanly typed code as much as possible.

I would also appreciate docstrings for methods which currently have none :-)...

Thanks

galexey committed:
fix for _is_my_own_echo and Eventtypes

galexey committed:
introduced a historylist of sent itemvalues for better echo checking

galexey committed:
changed behaviour for not yet implemented OH types

galexey committed:
introduced slotted sending of changes

galexey committed:
changed item Factory and item creation methods for slotted sending of changes

galexey committed:
openHAB3
added authentication to SSE through API token

galexey committed:
added widgets
more detailed logging again

galexey committed:
added widgets
more detailed logging again

galexey committed:
decoupled receiving of events from processing of events through an additional thread for processing(dispatching) of events.

galexey committed:
changes widget behaviour

galexey committed:
logging Q len

galexey committed:
token store
